c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +382,251 @@ Orphan<DynamicValue> JsonCodec::decode(
 
 // -----------------------------------------------------------------------------
 
+namespace _ {  // private
+
+class Parser {
+public:
+  Parser(kj::ArrayPtr<const char> input) : input_(input), remaining_(input_) {}
+  void parseValue(JsonValue::Builder &output) {
+    consumeWhitespace();
+    KJ_REQUIRE(remaining_.size() > 0, "JSON message ends prematurely.");
+
+    switch (nextChar()) {
+      case 'n': consume(NULL_); output.setNull();         break;
+      case 'f': consume(FALSE); output.setBoolean(false); break;
+      case 't': consume(TRUE);  output.setBoolean(true);  break;
+      case '"': parseString(output); break;
+      case '[': parseArray(output);  break;
+      case '{': parseObject(output); break;
+      // TODO(security): We could check for numbers more carefully instead of
+      // relying on strtod.
+      default: parseNumber(output); break;
+    }
+  }
+
+  void parseNumber(JsonValue::Builder &output) {
+    // TODO(someday): strtod allows leading +, while JSON grammar does not.
+    // strtod consumes leading whitespace, so we don't have to.
+    char *numEnd;
+    output.setNumber(std::strtod(remaining_.begin(), &numEnd));
+
+    advanceTo(numEnd);
+  }
+
+  void parseString(JsonValue::Builder &output) {
+    output.setString(consumeQuotedString());
+  }
+
+  void parseArray(JsonValue::Builder &output) {
+    // TODO(perf): Using orphans leaves holes in the message. It's expected
+    // that a JsonValue is used for interop, and won't be sent or written as a
+    // Cap'n Proto message.
+    kj::Vector<Orphan<JsonValue>> values;
+    auto orphanage = Orphanage::getForMessageContaining(output);
+
+    consume('[');
+    while (consumeWhitespace(), nextChar() != ']') {
+      auto orphan = orphanage.newOrphan<JsonValue>();
+      auto builder = orphan.get();
+      parseValue(builder);
+      values.add(kj::mv(orphan));
+
+      if (consumeWhitespace(), nextChar() != ']') {
+        // TODO(soon): This incorrectly allows a trailing comma.
+        consume(',');
+      }
+    }
+
+    output.initArray(values.size());
+    auto array = output.getArray();
+
+    for (auto i : kj::indices(values)) {
+      array.adoptWithCaveats(i, kj::mv(values[i]));
+    }
+
+    consume(']');
+  }
+
+  void parseObject(JsonValue::Builder &output) {
+    kj::Vector<Orphan<JsonValue::Field>> fields;
+    auto orphanage = Orphanage::getForMessageContaining(output);
+
+    consume('{');
+    while (consumeWhitespace(), nextChar() != '}') {
+      auto orphan = orphanage.newOrphan<JsonValue::Field>();
+      auto builder = orphan.get();
+
+      builder.setName(consumeQuotedString());
+
+      consumeWhitespace();
+      consume(':');
+      consumeWhitespace();
+
+      auto valueBuilder = builder.getValue();
+      parseValue(valueBuilder);
+
+      fields.add(kj::mv(orphan));
+
+      if (consumeWhitespace(), nextChar() != '}') {
+        // TODO(soon): This incorrectly allows a trailing comma.
```

Comment:
Well, if we're ever going to change it I think it should be now, but I don't think we ever need to change it so I'd make it a note rather than a TODO.